test/eckey.c

```diff
@@ -0,0 +1,132 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <openssl/conf.h>
+#include <openssl/ec.h>
+#include <openssl/engine.h>
+
+void setup(void)
+{
+    OPENSSL_load_builtin_modules();
+
+    ENGINE_load_builtin_engines();
+
+    /* CONF_MFLAGS_DEFAULT_SECTION introduced some time between 0.9.8b and
+       0.9.8e */
+#ifndef CONF_MFLAGS_DEFAULT_SECTION
+#define CONF_MFLAGS_DEFAULT_SECTION 0x0
+#endif
+
+    CONF_modules_load_file(NULL, NULL,
+                           CONF_MFLAGS_DEFAULT_SECTION|
+                           CONF_MFLAGS_IGNORE_MISSING_FILE);
+#if (OPENSSL_VERSION_NUMBER >= 0x10100000L) &&  \
+    !defined(LIBRESSL_VERSION_NUMBER)
+    /* OpenSSL 1.1.0+ takes care of initialization itself */
+#else
+    OpenSSL_add_all_algorithms();
+#endif
+}
+
+int check_eckey(int nid, const char *name)
+{
+    int            ret = 0;
+    ECDSA_SIG     *sig = NULL;
+    EC_KEY        *eckey = NULL;
+    unsigned char  digest[20];
+    ENGINE        *engine;
+
+    memset(digest, 0, sizeof(digest));
+
+    engine = ENGINE_by_id("ibmca");
+    if (engine == NULL) {
+        fprintf(stderr, "ibmca engine not loaded\n");
+        goto out;
+    }
+    eckey = EC_KEY_new_by_curve_name(nid);
+    if (eckey == NULL) {
+        /* curve not supported => test passed */
+        fprintf(stderr, "Curve %s not supported by OpenSSL\n", name);
+        ret = 1;
+        goto out;
+    }
+    if (EC_KEY_get0_engine(eckey) != engine) {
+        fprintf(stderr, "EC_KEY for %s does not use ibmca engine\n", name);
+        goto out;
+    }
+    if (!EC_KEY_generate_key(eckey)) {
+        /* error */
+        fprintf(stderr, "Failed to generate EC_KEY for %s\n", name);
+        goto out;
+    }
+    sig = ECDSA_do_sign(digest, sizeof(digest), eckey);
+    if (sig == NULL) {
+        /* error */
+        fprintf(stderr, "Failed to sign with %s\n", name);
+        goto out;
+    }
+    ret = ECDSA_do_verify(digest, sizeof(digest), sig, eckey);
+    if (ret == -1) {
+        /* error */
+        fprintf(stderr, "Failed to verify signature with %s\n", name);
+        goto out;
+    } else if (ret == 0) {
+        /* incorrect signature */
+        fprintf(stderr, "Signature incorrect with %s\n", name);
+        goto out;
+    } else {
+        /* signature ok */
+        ret = 1;
+    }
+ out:
+    if (engine)
+        ENGINE_free(engine);
+    if (sig)
+        ECDSA_SIG_free(sig);
+    if (eckey)
+        EC_KEY_free(eckey);
```

Comment:
All 3 functions check their arg for not null internally

Reply:
Could not find that in the documentation.  So I am not sure this is true also for old versions of openssl and would prefer to keep the check.

Reply:
Yes, I'm not sure for older openssl versions, I checked the code for v1.1.1.

Reply:
I also vote for keeping the check. I think its good practice to the initialize the heap pointers to NULL and free each at the exit label iff it is != NULL, so leaving the checks explicit makes the code more understandable and harder to mess up. If anything, even the engine pointer is initialized directly after the declarations, one could still initialize it to NULL, so in case order of memory allocation changes, the exit label code still holds. However, not a request to change it, I leave it up to you,

Reply:
Added the missing initialization.  Completely forgot this since this code was patched in later after the initial version...

Reply:
as i said it was not really "missing" for even if the allocation of heap items was changed afterwards, compiler would still issue a warning. more a question of personal taste. thanks anyway.